### PR TITLE
Add blackbox to workflow

### LINKED ISSUE #5261: 4.0.0-alpha.0 Blackbox tests
Add initial simple Blackbox tests for web3.js imports using web3 node module and minified build.

### PR DESCRIPTION
**Pull request by nikoulai**
## Description

Fixes #5261 

## Checklist:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `npm run dtslint` with success and extended the tests and types if necessary.
- [ ] I ran `npm run test:cov` and my test cases cover all the lines and branches of the added code.
- [ ] I ran `npm run build` with success.
- [ ] I have tested the built `dist/web3.min.js` in a browser.
- [ ] I have tested my code on the live network.
- [ ] I have checked the Deploy Preview and it looks correct.
- [ ] I have updated the `CHANGELOG.md` file in the root folder.

### CODE REVIEW

**Review by avkos — APPROVED** [review]
good job :muscle:

File: scripts/blackbox_test.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+# Exit immediately on error
+set -o errexit
+
+echo "Node version"
+node --version
+
+# Setup mock project to install web3 from virtual registry
+mkdir blackbox_test_app
+cp scripts/js/basic_usage.ts blackbox_test_app/basic_usage.ts
+cp -r fixtures blackbox_test_app/.
```
Comment (nazarhussain) [inline scripts/blackbox_test.sh]: I would suggest to keep one directory for the test app with all the files and `package.json` in it. And just copy that directory. 

This will be easy to extend that app with more behavior in it and cover more edge cases during the development.

File: verdaccio.yml [hunk 2]
```diff
@@ -0,0 +1,31 @@
+storage: ./storage
+auth:
+  htpasswd:
+    file: ./htpasswd
+uplinks:
```
Comment (nazarhussain) [inline verdaccio.yml:5]: Do we really need to define the uplinks?

File: verdaccio.yml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    publish: $all
+    proxy: npmjs yarn
+  '**':
+    access: $all
+    publish: $all
+    proxy: npmjs yarn
+logs:
+  - {type: stdout, format: pretty, level: warn}
+server:
+  keepAliveTimeout: 600 # specified in seconds
```
Comment (nazarhussain) [inline verdaccio.yml]:
```suggestion
  keepAliveTimeout: 600 # specified in seconds
  
```

File: scripts/npm.publish.sh [hunk 4]
```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env bash
```
Comment (nazarhussain) [inline scripts/npm.publish.sh]: The name of the file could be confusing, to actually publishing the scripts or publishing to local registry. May be rename it to `publish_ verdaccio.sh`

**Review by nazarhussain — CHANGES_REQUESTED** [review]

File: verdaccio.yml (hunk 2 above)
Comment (luu-alex) [reply verdaccio.yml:5]: I believe we do, otherwise we run into an error with packages not being found

File: scripts/blackbox_test.sh (hunk 1 above)
Comment (luu-alex) [reply scripts/blackbox_test.sh]: sounds good. created a`blackbox_test` folder

File: blackbox_test/basic_usage.ts [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+import util from 'util';
+import Web3 from 'web3';
+// import WebSocketProvider from 'web3-providers-ws';
```
Comment (Spacesai1or) [inline blackbox_test/basic_usage.ts]:
```suggestion
```
Remove commented import?

File: blackbox_test/basic_usage.ts [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+	let providerUrl = isWs(backendMode) ? process.env.INFURA_WSS : process.env.INFURA_HTTP;
+
+	if (providerUrl === undefined) throw new Error('Provider url is undefined!');
+	else providerUrl = providerUrl as string;
+
+	log('Running on backend mode: ', backendMode);
+
+	web3 = new Web3(providerUrl);
+
+	if (isWs(backendMode)) console.log('^^^^^^^^^', await waitForOpenConnection(web3, backendMode));
```
Comment (Spacesai1or) [inline blackbox_test/basic_usage.ts]:
```suggestion
	if (isWs(backendMode)) log('^^^^^^^^^', await waitForOpenConnection(web3, backendMode));
```

File: blackbox_test/basic_usage.ts [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+				clearInterval(interval);
+				resolve();
+			}
+			// eslint-disable-next-line no-plusplus, no-param-reassign
+			currentAttempt++;
+		}, intervalTime);
+	});
+const { log } = console;
+
+async function main() {
```
Comment (Spacesai1or) [inline blackbox_test/basic_usage.ts:56]: Shouldn't this be making use of a test framework so that we can `expect` the results we're getting back instead of just using `util.inspect`?

**Review by nazarhussain — APPROVED** [review]
Apart from @spacesailor24  comments, LGTM.

**Review by Spacesai1or — CHANGES_REQUESTED** [review]
Maybe I'm misunderstanding something, but it appears that the `yarn:test:blackbox:web` tests [stop](https://github.com/ChainSafe/web3.js/runs/7789883394?check_suite_focus=true#step:4:1) after making the minified build and don't seem to be running any tests

**Comment by luu-alex** [comment]
@spacesailor24 i wasnt sure if web was complete or not, but its related to the html file that was created. `blackbox_test/index.html`

**Review by jdevcs — CHANGES_REQUESTED** [review]
> Maybe I'm misunderstanding something, but it appears that the `yarn:test:blackbox:web` tests [stop](https://github.com/ChainSafe/web3.js/runs/7789883394?check_suite_focus=true#step:4:1) after making the minified build and don't seem to be running any tests

For blackbox test of using web3.js in js script file we can see logs manually now like https://github.com/ChainSafe/web3.js/runs/7789853708?check_suite_focus=true#step:4:2586

For black box web , you are right , can we show logs at least in CI action if any call in blackbox-web is failing? or for now we can check it manually on local sys.   @spacesailor24

**Review by jdevcs — CHANGES_REQUESTED** [review]
This needs to be fixed:
https://github.com/ChainSafe/web3.js/issues/5345

File: blackbox_test/basic_usage.ts [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+	if (providerUrl === undefined) throw new Error('Provider url is undefined!');
+	else providerUrl = providerUrl as string;
+
+	log('Running on backend mode: ', backendMode);
+
+	web3 = new Web3(providerUrl);
+
+	if (isWs(backendMode)) log('^^^^^^^^^', await waitForOpenConnection(web3, backendMode));
+
+	log();
```
Comment (jdevcs) [inline blackbox_test/basic_usage.ts:74]: also log `Web3.version` here

**Review by jdevcs — CHANGES_REQUESTED** [review]

**Comment by jdevcs** [comment]
@spacesailor24 @nikoulai

**Comment by jdevcs** [comment]
Hi @nikoulai can we close this PR for https://github.com/ChainSafe/web3.js/pull/5353
